Configure dependabot to automated version updates

daniel-zullo-frequenz opened the pull request:
Add dependabot configuration file to automate version updates for pip dependencies and github actions.

The security updates needs to be checked/enabled from the settings in the repo.

Fixes: #242

daniel-zullo-frequenz committed:
Configure dependabot version updates for pip dependencies

Signed-off-by: daniel-zullo-frequenz <120166726+daniel-zullo-frequenz@users.noreply.github.com>

daniel-zullo-frequenz committed:
Configure dependabot version updates for github actions

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

leandro-lucarella-frequenz commented on .github/dependabot.yml:
```diff
@@ -0,0 +1,22 @@
+    schedule:
+      interval: "weekly"
+    labels:
+      - "part:tooling"
```
Can you add `type:tech-debt` too (for both)?

daniel-zullo-frequenz replied:
sure, I was not sure about which other labels to add

leandro-lucarella-frequenz replied:
I think we are good with these 2. There is also prio, but I guess for now it is fine to leave it without, unless is a security update, if it is maybe `prio:high` would be a good.

daniel-zullo-frequenz replied:
I'm afraid we can only set `priority:high` for security updates if we define a dedicated update for the default branch as it was done for the supported branches.

leandro-lucarella-frequenz replied:
I see, and do you know if there is no way to use wildcards for `target-branch` then? I had a quick look at the docs and it doesn't say, I wonder if you tried it.

Also the docs says that `target-branch` doesn't apply to security updates, so I'm not sure how this will play along.

~BTW, using [YAML anchors](https://community.atlassian.com/t5/Bitbucket-questions/YAML-anchors-and-variables-code-reuse-question/qaq-p/1733738) we can remove **a lot** of duplication.~ But anyway, if we are going to need to keep this file updated with each release, I think it will get out of sync pretty soon, so I'm not sure it is worth adding the `target-branch` if there is no wildcard support (at least until we include updating this file in the release automation).

Had a more in depth look and looking at the dependabot code, no, it doesn't support wildcards, [is doing just a simple comparison](https://github.com/dependabot/dependabot-core/blob/e64b21e4304b98570a585b8ed6fd74cd90142cfb/common/lib/dependabot/config/file.rb#L21).

I looked for issues and only found one about [supporting arrays for `target-branch`](https://github.com/dependabot/dependabot-core/issues/2511), but that doesn't help us, so [I created one for wildcards support](https://github.com/dependabot/dependabot-core/issues/6890).

leandro-lucarella-frequenz replied:
I'd say let's ignore other branches for now until things are more stable, we have better automation, or the feature is added. At the most we could add a `stable` branch to point to the current release and add a dependabot config only for that one. But I would skip that for now too, let's first try this out.

leandro-lucarella-frequenz commented on .github/dependabot.yml:
```diff
@@ -0,0 +1,22 @@
+  - package-ecosystem: "github-actions"
+    directory: "/"
+    schedule:
+      interval: "weekly"
```
Why one weekly and the other daily?

daniel-zullo-frequenz replied:
I don't have a specific reason for it, but I think that checking the versions of  Github Actions on a weekly basis would be enough.
Since we have many pip dependencies, Dependabot creates a pull request for every version update. By default, only up to 5 PRs can be open at a time (although I've adjusted the configuration to allow up to 10). If we update the dependencies versions on a daily basis, it may help ensure that they stay up-to-date in case we reach the PR limit (and some PRs get merged).

leandro-lucarella-frequenz replied:
These can also have security issues, which could leak credentials for example. At least for security issues I would check it daily.

daniel-zullo-frequenz replied:
That's a good point, `daily` it is then.

leandro-lucarella-frequenz commented on .github/dependabot.yml:
```diff
@@ -0,0 +1,22 @@
+    schedule:
+      interval: "daily"
+      time: "07:00"
+    versioning-strategy: auto
```
Maybe add a comment about this? Is this for how to interpret dependencies versions or if to update to minor or major versions? Or for how the project version itself? Maybe add a comment (with a link to the docs)?

daniel-zullo-frequenz replied:
This is the default versioning-strategy. For apps, the version requirements are increased. For libraries, the range of versions is widened. It isn't required but I added explicitly in case we need to change it. I'll add a reference to the docs

leandro-lucarella-frequenz replied:
Maybe add a comment saying just that? (and add the link to the docs?)

leandro-lucarella-frequenz replied:
Also, how does it know that a repo is an app or a lib? No dependencies lock file means is a lib?

daniel-zullo-frequenz replied:
It isn't precisely specified in the documentation as it depends on the ecosystem. My understanding is it depends on an entry point in the manifest file(s) to consider the repo an app

leandro-lucarella-frequenz reviewed (commented):
Do you have any examples of how this work in your fork to understand better how it works?

A few other questions:

1. Will this propose only security updates or will also propose updates when any dependency gets a new version released?
2. If it is the second, will propose updates only for minor updates? Major too? IIRC there were options to control this.
3. Will this run on all branches or just the default branch? If the second, can it be enabled to older versions also can get security updates?

daniel-zullo-frequenz commented:
> Do you have any examples of how this work in your fork to understand better how it works?

Yes, you can see some PR examples [here](https://github.com/daniel-zullo-frequenz/frequenz-sdk-python/pulls?q=is%3Apr+is%3Aopen+sort%3Aupdated-desc)

daniel-zullo-frequenz commented:
> Will this propose only security updates or will also propose updates when any dependency gets a new version released?

The securities update are managed/configured in the settings of the repo, see image below. We only need to ensure/check security updates are enabled

![image](https://user-images.githubusercontent.com/120166726/225873275-96fe768a-f15c-4ac8-87b7-4573c4184e06.png)

The configuration file will take care only of version updates

daniel-zullo-frequenz commented:
> Will this run on all branches or just the default branch? If the second, can it be enabled to older versions also can get security updates?

The current configuration will run only on the default branch. Though we can set manually another branch using `target-branch`

daniel-zullo-frequenz commented:
Added the supported branches so that security updates are automatically applied on those branches. Unfortunately I couldn't use anchor and aliases in the YAML for dependabot as they not supported due to security issues with the yaml parser for some ecosystems. Thus you might notice there are code duplication/redundancy in the configuration

leandro-lucarella-frequenz commented:
> The securities update are managed/configured in the settings of the repo, see image below. We only need to ensure/check security updates are enabled

I see, so we have no control over security updates? Like frequency, labels, etc? I thought there was some config possible with the yaml file.

leandro-lucarella-frequenz commented:
> The current configuration will run only on the default branch. Though we can set manually another branch using target-branch

Do you know if we can use wildcards? I would enable it for all branches. We can just remove unmaintained branches so they don't get PRs, this will clearly indicate which branches are maintained and which aren't anymore.

leandro-lucarella-frequenz commented:
> Yes, you can see some PR examples [here](https://github.com/daniel-zullo-frequenz/frequenz-sdk-python/pulls?q=is%3Apr+is%3Aopen+sort%3Aupdated-desc)

A few comments:

* I see all those PRs have `prio:urgent`, is this because you configured it like this in your test repo?

* So mainly all updates are breaking except 2 right? I see major versions are being updated. I repeat the question, can we configure if we want to do patch, minor or major updates, or all? If all, is it possible to have 3 separate PRs, as the patch and minor should normally succeed and the major break?

  OK, actually for libs I guess patch and minor updates makes no sense, and we'll test them anyway and is only the minor requirements, so we probably don't want to bump them unnecessarily. But for apps it probably makes sense to try all of them (and separately).

* Is it possible to configure unstable versions (0.x.x) to consider minor updates as breaking too, so as major updates? Or if this is the default?

daniel-zullo-frequenz commented:
> Do you know if we can use wildcards? I would enable it for all branches. We can just remove unmaintained branches so they don't get PRs, this will clearly indicate which branches are maintained and which aren't anymore.

I'm afraid we cannot use wildcards either in the current state :(
There are feature-requests to add support for yaml aliases and wildcards, or support multiple target-branches given a one update-configuration but at the moment we might need to live with the code duplication if we use dependabot

daniel-zullo-frequenz commented:
> I see all those PRs have prio:urgent, is this because you configured it like this in your test repo?

No, that was because I accidentally set version updates for stable branches (non-default) instead of only security updates.

> Is it possible to configure unstable versions (0.x.x) to consider minor updates as breaking too, so as major updates? Or if this is the default?

I'm afraid we cannot configure that or I haven´t found it yet how to do it. Anyway let's please discuss it in a call

leandro-lucarella-frequenz commented:
OK, we met and here is the summary of what we talked about:

* There is a lot of documentation lacking for dependabot, which makes it hard to know how it will work. This might be a reason to look for alternatives eventually, but we'll still try dependabot for now because of the easy integration.
* About libraries vs apps, we couldn't find any source of truth, but we ended up assuming that if a dependency is specified as a range, then probably dependabot will just widen the dependency and if it is pinned, it will update the pinning.
* We decided to pin all dev dependencies to a particular version, as we don't really need to support a range. This will maximize the reproducibility of builds and dependabot will let us know when there are new versions of the dependencies so we can easily bump them.
* We decide to remove all hacks to support branches other the default branch. For now we only get updates in the default branch and we'll manually backport anything that we deem important.
* The configuration file is only for maintenance bumping of dependencies. Security updates are handled in the repo config (it is already enabled) and can't be configured any further.

daniel-zullo-frequenz committed:
Pin dev/test dependencies to a specific version

The SDK does not really need to support a range of version for
dev/test dependencies. This will maximize the reproducibility
of builds and dependabot will open PRs to let us know when there
are new versions of the dependencies so that we can easily bump them.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz commented:
Thanks Luca for the nice write up!
I've updated the PR to remove the security updates hack on supported branches and to pin dev/test dependencies to a particular version.

leandro-lucarella-frequenz reviewed (dismissed):
Excellent!

Last comment (optional): We don't really need to set the `reviewers` here, as we already have it in the `CODEOWNERS` file, it's sort of redundant.

daniel-zullo-frequenz commented:
> We don't really need to set the reviewers here, as we already have it in the CODEOWNERS file, it's sort of redundant.

I agree and I've already removed reviewers.
